File: plugins/plugin-twitch/rust/src/providers/channel_state.rs
//! Channel state provider for Twitch plugin.

use crate::service::TwitchService;
use crate::types::{format_channel_for_display, normalize_channel};
use serde::{Deserialize, Serialize};
use std::sync::Arc;

/// Channel state data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChannelStateData {
    pub channel: String,
    pub display_channel: String,
    pub is_primary_channel: bool,
    pub bot_username: String,
    pub joined_channels: Vec<String>,
    pub channel_count: usize,
    pub connected: bool,
}

/// Channel state values (simple key-value)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChannelStateValues {
    pub channel: String,
    pub display_channel: String,
    pub is_primary_channel: bool,
    pub bot_username: String,
    pub channel_count: usize,
}

/// Channel state provider result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChannelStateResult {
    pub data: ChannelStateData,
    pub values: ChannelStateValues,
    pub text: String,
}

/// Get the current Twitch channel state
pub async fn get_channel_state(
    service: Arc<TwitchService>,
    channel: Option<&str>,
    agent_name: &str,
) -> ChannelStateResult {
    let channel = channel
        .map(normalize_channel)
        .unwrap_or_else(|| normalize_channel(service.get_primary_channel()));

    let joined_channels = service.get_joined_channels().await;
    let is_primary_channel = channel == normalize_channel(service.get_primary_channel());
    let bot_username = service.get_bot_username().to_string();
    let display_channel = format_channel_for_display(&channel);
    let connected = service.is_connected().await;

    let mut text = format!(
        "{} is currently in Twitch channel {}.",
        agent_name, display_channel
    );

    if is_primary_channel {
        text.push_str(" This is the primary channel.");
    }

    text.push_str("\n\nTwitch is a live streaming platform. Chat messages are public and visible to all viewers.");
    text.push_str(&format!(" {} is logged in as @{}.", agent_name, bot_username));
    text.push_str(&format!(
        " Currently connected to {} channel(s).",
        joined_channels.len()
    ));

    ChannelStateResult {
        data: ChannelStateData {
            channel: channel.clone(),
            display_channel: display_channel.clone(),
            is_primary_channel,
            bot_username: bot_username.clone(),
            joined_channels,
            channel_count: service.get_joined_channels().await.len(),
            connected,
        },
        values: ChannelStateValues {
            channel,
            display_channel,
            is_primary_channel,
            bot_username,
            channel_count: service.get_joined_channels().await.len(),
        },
        text,
    }
}

/// Provider metadata
pub const PROVIDER_NAME: &str = "twitchChannelState";
pub const PROVIDER_DESCRIPTION: &str = "Provides information about the current Twitch channel context";

File: plugins/plugin-twitch/typescript/src/providers/channelState.ts
/**
 * Channel state provider for Twitch plugin.
 */

import type {
  IAgentRuntime,
  Memory,
  Provider,
  ProviderResult,
  State,
} from "@elizaos/core";
import type { TwitchService } from "../service.js";
import {
  formatChannelForDisplay,
  normalizeChannel,
  TWITCH_SERVICE_NAME,
} from "../types.js";

/**
 * Provider that gives the agent information about the current Twitch channel context.
 */
export const channelStateProvider: Provider = {
  name: "twitchChannelState",
  description: "Provides information about the current Twitch channel context",

  get: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State,
  ): Promise<ProviderResult> => {
    // Only provide context for Twitch messages
    if (message.content.source !== "twitch") {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const twitchService =
      await runtime.getService<TwitchService>(TWITCH_SERVICE_NAME);

    if (!twitchService || !twitchService.isConnected()) {
      return {
        data: {
          connected: false,
        },
        values: {
          connected: false,
        },
        text: "",
      };
    }

    const agentName = state?.agentName || "The agent";

    // Get room from state if available
    const room = state?.data?.room;
    const channelId = room?.channelId as string | undefined;
    const channel = channelId
      ? normalizeChannel(channelId)
      : twitchService.getPrimaryChannel();

    const joinedChannels = twitchService.getJoinedChannels();
    const isPrimaryChannel = channel === twitchService.getPrimaryChannel();
    const botUsername = twitchService.getBotUsername();

    let responseText = `${agentName} is currently in Twitch channel ${formatChannelForDisplay(channel)}.`;

    if (isPrimaryChannel) {
      responseText += " This is the primary channel.";
    }

    responseText += `\n\nTwitch is a live streaming platform. Chat messages are public and visible to all viewers.`;
    responseText += ` ${agentName} is logged in as @${botUsername}.`;
    responseText += ` Currently connected to ${joinedChannels.length} channel(s).`;

    return {
      data: {
        channel,
        displayChannel: formatChannelForDisplay(channel),
        isPrimaryChannel,
        botUsername,
        joinedChannels,
        channelCount: joinedChannels.length,
        connected: true,
      },
      values: {
        channel,
        displayChannel: formatChannelForDisplay(channel),
        isPrimaryChannel,
        botUsername,
        channelCount: joinedChannels.length,
      },
      text: responseText,
    };
  },
};
